RecoHI/HiEvtPlaneAlgos/src/EvtPlaneProducer.cc

```diff
@@ -305,214 +498,207 @@ void EvtPlaneProducer::produce(edm::Event &iEvent, const edm::EventSetup &iSetup
-            if (MomConsWeight[i][0] == 'y' && loadDB_) {
-              w = flat[i]->getW(tower_energyet, vzr_sell, bin);
-            }
-            if (tower_eta < 0)
-              w = -w;
-          }
-          rp[i]->addParticle(w, tower_energyet, sin(EPOrder[i] * tower_phi), cos(EPOrder[i] * tower_phi), tower_eta);
-        }
+    iEvent.getByToken(lostToken, cands);
+    iEvent.getByToken(chi2MapLostToken, chi2Map);
+    for (unsigned int i = 0, n = cands->size(); i < n; ++i) {
+      track = {};
+      track.centbin = cbin;
+      if (cuts->isGoodHF(track)) {
+        FillHF(track, bestvz, bin);
+      }
+      const pat::PackedCandidate &pf = (*cands)[i];
+      if (!pf.hasTrackDetails())
+        continue;
+      const reco::Track &trk = pf.pseudoTrack();
+      track.highPurity = pf.trackHighPurity();
+      track.charge = trk.charge();
+      if (!track.highPurity || track.charge == 0)
+        continue;
+      track.collection = 2;
+      track.eta = pf.eta();
+      track.phi = pf.phi();
+      track.et = pf.et();
+      track.pdgid = pf.pdgId();
+      track.pt = trk.pt();
+      track.ptError = trk.ptError();
+      track.numberOfValidHits = trk.numberOfValidHits();
+      track.algos = trk.algo();
+      track.dz = trk.dz(bestvtx);
+      track.dxy = -1. * trk.dxy(bestvtx);
+      track.dzError = sqrt(pow(trk.dzError(), 2) + pow(bestvzError, 2));
+      track.dxyError = sqrt(pow(trk.dxyError(), 2) + bestvxError * bestvyError);
+      track.dzSig = track.dz / track.dzError;
+      track.dxySig = track.dxy / track.dxyError;
+      const reco::HitPattern &hit_pattern = trk.hitPattern();
+      track.normalizedChi2 = (*chi2Map)[cands->ptrAt(i)];
+      track.chi2layer = (*chi2Map)[cands->ptrAt(i)] / hit_pattern.trackerLayersWithMeasurement();
+      if (cuts->isGoodTrack(track)) {
+        FillTracker(track, bestvz, bin);
       }
     }
-  }
 
-  //Castor part
-
-  iEvent.getByToken(castorToken, castorCollection_);
-
-  if (castorCollection_.isValid()) {
-    for (std::vector<reco::CastorTower>::const_iterator j = castorCollection_->begin(); j != castorCollection_->end();
-         j++) {
-      tower_eta = j->eta();
-      tower_phi = j->phi();
-      tower_energyet = j->et();
-      double minet = minet_;
-      double maxet = maxet_;
-      for (int i = 0; i < NumEPNames; i++) {
-        if (EPDet[i] == Castor) {
-          if (minet_ < 0)
-            minet = minTransverse[i];
-          if (maxet_ < 0)
-            maxet = maxTransverse[i];
-          if (tower_energyet < minet)
-            continue;
-          if (tower_energyet > maxet)
-            continue;
-          double w = tower_energyet;
-          if (EPOrder[i] == 1) {
-            if (MomConsWeight[i][0] == 'y' && loadDB_) {
-              w = flat[i]->getW(tower_energyet, vzr_sell, bin);
-            }
-            if (tower_eta < 0)
-              w = -w;
+  } else {
+    //calorimetry part
+    if (scalo.find("particleFlow") != std::string::npos) {
```

Comment:
compute a boolean flag in the constructor and use in event by event, no need to check the string repeatedly